src/lib/Components/Inbox/Conversations/Messages.tsx

```diff
@@ -0,0 +1,176 @@
+            onSelected={() => ARSwitchBoard.presentNavigationViewController(this, conversationItem.href)}
+          />
+        }
+        showPreview={
+          item.first_message &&
+          conversationItem.__typename === "Show" &&
+          <ShowPreview
+            show={conversationItem}
+            onSelected={() => ARSwitchBoard.presentNavigationViewController(this, conversationItem.href)}
+          />
+        }
+      />
+    )
+  }
+
+  loadMore() {
+    // if (!this.props.relay.hasMore() || this.props.relay.isLoading()) {
+    //   return
+    // }
+
+    // TODO show loading indicator
+
+    this.props.relay.loadMore(10, e => {
+      // TODO hide loading indicator
+      console.log(e)
+    })
+  }
+
+  render() {
+    const messages = (this.props.conversation.messages || { edges: [] }).edges.map((edge, index) => {
+      return { first_message: index === 0, key: index, ...edge.node }
+    })
+    return (
+      <FlatList
+        data={messages}
+        renderItem={this.renderMessage.bind(this)}
+        onEndReached={this.loadMore.bind(this)}
+        onEndReachedThreshold={0.2}
+        ItemSeparatorComponent={DottedBorder}
+      />
+    )
+  }
+}
+
+export default createPaginationContainer(
+  Messages,
+  {
+    conversation: graphql.experimental`
+      fragment Messages_conversation on Conversation {
+        __id
+        id
+        from {
+          name
+          email
+          initials
+        }
+        to {
+          name
+          initials
+        }
+        messages(first: $count, after: $after) @connection(key: "Messages_messages") {
+          pageInfo {
+            startCursor
+            endCursor
+            hasPreviousPage
+            hasNextPage
+          }
+          edges {
+            cursor
+            node {
+              impulse_id
+              is_from_user
+              body
+              ...Message_message
+            }
+          }
+        }
+        items {
+          item {
+            ... on Artwork {
+              __typename
+              href
+              ...ArtworkPreview_artwork
+            }
+            ... on Show {
+              __typename
+              href
+              ...ShowPreview_show
+            }
+          }
+        }
+      }
+    `,
+  },
+  {
+    direction: "forward",
+    getConnectionFromProps(props) {
+      return props.conversation && props.conversation.messages
+    },
+    getFragmentVariables(prevVars, totalCount) {
+      return {
+        ...prevVars,
+        count: totalCount,
+      }
+    },
+    getVariables(props, paginationInfo, fragmentVariables) {
+      return {
+        conversationID: props.conversation.id,
+        count: paginationInfo.count,
+        after: paginationInfo.cursor,
+      }
+    },
+    query: graphql`
+      query MessagesPaginationQuery($conversationID: String!, $count: Int!, $after: String) {
+        me {
+          conversation(id: $conversationID) {
+            ...Messages_conversation
```

Comment:
You probably should use `@arguments` here and add `@argumentDefinitions` in the fragment above, at least that’s what I’ve been doing, still not entirely sure when that’s needed.

Reply:
Ohhh, only now see your comments.... sorry for piling on

Reply:
it is needed for local variables definitions

Reply:
@sibelius Right, and so it scopes them to just these fragments? Looking at the docs, it seems like they indicate that it should be possible to do without the experimental features, is that really true? Because I had a hard time trying that.

Reply:
graphql.experimental is deprecated on master already

Reply:
Makes sense, good to know! 👍